superset-frontend/src/hooks/apiResources/schemas.ts

```diff
@@ -16,65 +16,97 @@
-  return SupersetClient.get({ endpoint }) as Promise<QueryData>;
-}
+export type FetchSchemasQueryParams = {
+  dbId?: string | number;
+  forceRefresh: boolean;
+  onSuccess?: (data: SchemaOption[], isRefetched: boolean) => void;
+  onError?: () => void;
+};
+
+type Params = Omit<FetchSchemasQueryParams, 'forceRefresh'>;
+
+const schemaApi = api.injectEndpoints({
+  endpoints: builder => ({
+    schemas: builder.query<SchemaOption[], FetchSchemasQueryParams>({
+      providesTags: [{ type: 'Schemas', id: 'LIST' }],
+      query: ({ dbId, forceRefresh }) => ({
+        endpoint: `/api/v1/database/${dbId}/schemas/`,
+        // TODO: Would be nice to add pagination in a follow-up. Needs endpoint changes.
+        urlParams: {
+          force: forceRefresh,
+        },
+        transformResponse: ({ json }: JsonResponse) =>
+          json.result.map((value: string) => ({
+            value,
+            label: value,
+            title: value,
+          })),
+      }),
+      serializeQueryArgs: ({ queryArgs: { dbId } }) => ({
+        dbId,
+      }),
+    }),
+  }),
+});
 
-type Params = FetchSchemasQueryParams &
-  Pick<UseQueryOptions<SchemaOption[]>, 'onSuccess' | 'onError'>;
+export const { useLazySchemasQuery, useSchemasQuery } = schemaApi;
+
+const EMPTY_SCHEMAS = [] as SchemaOption[];
 
 export function useSchemas(options: Params) {
+  const isMountedRef = useRef(false);
   const { dbId, onSuccess, onError } = options || {};
-  const forceRefreshRef = useRef(false);
-  const params = { dbId };
-  const result = useQuery<QueryData, Error, SchemaOption[]>(
-    ['schemas', { dbId }],
-    () => fetchSchemas({ ...params, forceRefresh: forceRefreshRef.current }),
+  const [trigger, refetchResult] = useLazySchemasQuery();
+  const result = useSchemasQuery(
+    { dbId, forceRefresh: false },
     {
-      select: ({ json }) =>
-        json.result.map((value: string) => ({
-          value,
-          label: value,
-          title: value,
-        })),
-      enabled: Boolean(dbId),
-      onSuccess,
-      onError,
-      onSettled: () => {
-        forceRefreshRef.current = false;
-      },
+      skip: !dbId,
     },
   );
 
+  const refetch = useCallback(() => {
+    if (dbId) {
+      trigger({ dbId, forceRefresh: true }).then(
+        ({ isSuccess, isError, data }) => {
+          if (isSuccess) {
+            onSuccess?.(data || EMPTY_SCHEMAS, true);
+          }
+          if (isError) {
+            onError?.();
+          }
+        },
+      );
+    }
+  }, [dbId]);
+
+  useEffect(() => {
+    if (isMountedRef.current) {
+      const { requestId, isSuccess, isError, isFetching, data, originalArgs } =
+        result;
+      if (!originalArgs?.forceRefresh && requestId && !isFetching) {
+        if (isSuccess) {
+          onSuccess?.(data || EMPTY_SCHEMAS, false);
+        }
+        if (isError) {
+          onError?.();
+        }
+      }
+    } else {
+      isMountedRef.current = true;
+    }
+  }, [result]);
```

Comment:
Can you resolve the dependency warnings of the effects?

Reply:
waiting #23871 to clean up dependency warning